test/performance/odp_packet_gen.c

```diff
@@ -0,0 +1,1232 @@
+			ret--;
+		}
+	}
+
+	odp_barrier_init(&global->barrier, num_cpu + 1);
+
+	return 0;
+}
+
+static int open_pktios(test_global_t *global)
+{
+	odp_pool_capability_t pool_capa;
+	odp_pktio_capability_t pktio_capa;
+	odp_pool_param_t  pool_param;
+	odp_pool_t pool;
+	odp_pktio_param_t pktio_param;
+	odp_pktio_t pktio;
+	odp_pktio_config_t pktio_config;
+	odp_pktin_queue_param_t pktin_param;
+	odp_pktout_queue_param_t pktout_param;
+	char *name;
+	uint32_t i;
+	int j;
+	test_options_t *test_options = &global->test_options;
+	uint32_t num_rx = test_options->num_rx;
+	int num_tx = test_options->num_tx;
+	uint32_t num_pktio = test_options->num_pktio;
+	uint32_t num_pkt = test_options->num_pkt;
+	uint32_t pkt_len = test_options->pkt_len;
+	odp_pktout_queue_t pktout[num_tx];
+
+	printf("\nODP packet generator\n");
+	printf("  quit test after     %" PRIu64 " rounds\n",
+	       test_options->quit);
+	printf("  num rx threads      %u\n", num_rx);
+	printf("  num tx threads      %i\n", num_tx);
+	printf("  num packets         %u\n", num_pkt);
+	printf("  packet length       %u bytes\n", pkt_len);
+	printf("  tx burst size       %u\n", test_options->burst_size);
+	printf("  tx bursts           %u\n", test_options->bursts);
+	printf("  tx burst gap        %" PRIu64 " nsec\n",
+	       test_options->gap_nsec);
+	printf("  clock resolition    %" PRIu64 " hz\n", odp_time_local_res());
+	printf("  IPv4 source         %s\n", test_options->ipv4_src_s);
+	printf("  IPv4 destination    %s\n", test_options->ipv4_dst_s);
+	printf("  UDP source          %u\n", test_options->udp_src);
+	printf("  UDP destination     %u\n", test_options->udp_dst);
+	printf("  UDP src count       %u\n", test_options->c_mode.udp_src);
+	printf("  UDP dst count       %u\n", test_options->c_mode.udp_dst);
+	printf("  num pktio           %u\n", num_pktio);
+
+	printf("  interfaces names:   ");
+	for (i = 0; i < num_pktio; i++) {
+		if (i > 0)
+			printf("                      ");
+		printf("%s\n", test_options->pktio_name[i]);
+	}
+
+	printf("  destination MACs:   ");
+	for (i = 0; i < num_pktio; i++) {
+		uint8_t *eth_dst = global->pktio[i].eth_dst.addr;
+
+		if (i > 0)
+			printf("                      ");
+		printf("%02x:%02x:%02x:%02x:%02x:%02x\n",
+		       eth_dst[0], eth_dst[1], eth_dst[2],
+		       eth_dst[3], eth_dst[4], eth_dst[5]);
+	}
+	printf("\n");
+
+	global->pool = ODP_POOL_INVALID;
+
+	if (odp_pool_capability(&pool_capa)) {
+		printf("Error: Pool capability failed.\n");
+		return -1;
+	}
+
+	if (num_pkt > pool_capa.pkt.max_num) {
+		printf("Error: Too many packets. Max %u supported.\n",
+		       pool_capa.pkt.max_num);
+		return -1;
+	}
+
+	if (pkt_len > pool_capa.pkt.max_len) {
+		printf("Error: Too large packets. Max %u supported length.\n",
+		       pool_capa.pkt.max_len);
+		return -1;
+	}
+
+	/* Create pool */
+	odp_pool_param_init(&pool_param);
+	pool_param.type    = ODP_POOL_PACKET;
+	pool_param.pkt.num = num_pkt;
+	pool_param.pkt.len = pkt_len;
+
```

Comment:
Set `pool_param.pkt.seg_len` accordingly to guarantee that the packet header fits into the first segment.